lib/Core/Memory.cpp

```diff
@@ -901,15 +994,34 @@ ref<Expr> PersistentState::ptrAsExpr(void *ptr) {
   return ConstantExpr::create((uint64_t)ptr, Expr::Int64);
 }
 
-std::string PersistentState::getLocationInfo(const ExecutionState &state) {
+ref<ConstantExpr> 
+PersistentState::createRootCauseIdExpr(const ExecutionState &state, 
+                                       ref<Expr> cacheLineOffset,
+                                       const char *type) {
+
+  std::string info = getLocationInfo(state, cacheLineOffset, type);
+  if (!allRootLocations[info]) {
+    allRootLocations[info] = nextLocId;
+    nextLocId++;
+    assert(nextLocId <= UINT64_MAX && "run out of causes!");
+  }
+
+  return ConstantExpr::create(allRootLocations[info], rootCauseWidth);                               
+}
+
+std::string PersistentState::getLocationInfo(const ExecutionState &state, 
+                                             ref<Expr> offset, 
+                                             const char *type) const {
   const InstructionInfo *iip = state.prevPC->info;
 
   std::string infoStr;
   llvm::raw_string_ostream msg(infoStr);
   msg << "Persistent Memory Info:\n";
   msg << "\tName: " << getObject()->name << "\n";
-  msg << "\tAddress: " << getObject()->address << "\n";
+  msg << "\tAddress: " << llvm::format_hex(getObject()->address, 18) << "\n";
+  msg << "\t\tOffset:" << *offset << "\n";
```

Comment:
Now we will end up with many more unique root causes. I think we can do this better:

```
struct RootCause {
  enum Type {
    Write,
    Flush
  } type;
  const InstructionInfo *instInfo;
  mutable std::vector<ref<Expr>> offsetsInvokedWith;
  // or
  mutable ref<Expr> mostRecentOffsetInvokedWith;

  size_t hashVal;

  RootCause(const ExecutionState &state, ref<Expr> offset, Type type) {
    this->type = type;
    this->mostRecentOffsetInvokedWith = offset;
    this->instInfo = state.prevPC->info;

    // only hash using the bits we care about differentiating between
    std::string hashStr;
    llvm::raw_string_ostream msg(hashStr);
    msg << type;
    state.dumpStack(msg);
    this->hashVal = std::hash<std::string>(hashStr);
  }
};

```

Ideally, a root cause is one line of code. With this approach, we will report every unique _execution_ of a single line of code. At worst, only 1/64 of our reported bugs is unique.

Reply:
I realize the above claim is not correct; we will only report whichever root cause the solver finds to be stored at that cache line.

But still, this is a lot of strings to be storing.

Reply:
In general, I have realized the need to fix this up. Our root causes should really just be the source locations, with other information being separate (like the object, etc). I will fix this later.

Reply:
I can also do something with it if you'd like